### PR TITLE
Slashing protection pruning

### PR DESCRIPTION
**Pull request by jframe**
Add slashing protection pruning. This will run on a periodic process to remove any data in the slashing protection database that is below the specified threshold.

It is disabled by default.

The following new parameters have been added as part of pruning
* --slashing-protection-pruning-enabled to enable the pruning
* --slashing-protection-pruning-epochs-to-keep Number of epochs to keep
* --slashing-protection-pruning-slots-per-epoch Slots per epoch to use when calculating the number of slots to prune for signed blocks
* --slashing-protection-pruning-interval Hours between pruning operations

### CODE REVIEW

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2SubCommand.java [hunk 1]
```diff
@@ -129,6 +111,14 @@ protected void validateArgs() {
     }
   }
 
+  private void validatePositiveValue(final long value, final String fieldName) {
+    if (value < 1) {
```
Comment (usmansaleem) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2SubCommand.java:115]: `value <= 1`

**Review by usmansaleem — APPROVED** [review]
LGTM

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2SubCommand.java (hunk 1 above)
Comment (jframe) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2SubCommand.java:115]: If I change it to that then this won't allow 1 as a value. I only want to exclude zero and less than zero values as these don't make sense for these options.

File: slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import org.apache.tuweni.units.bigints.UInt64;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+public class PruningIntegrationTest extends IntegrationTestBase {
+
+  @ParameterizedTest
+  @CsvSource({"1, 9", "2, 8", "5, 5", "9, 1", "10, 0", "20, 0"})
+  void prunesDataForRegisteredValidator(final int amountToKeep, final int pruningMark) {
```
Comment (rain-on) [inline slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java:33]: do you need both numbers as parameters - it appears 90% of the time they add to 10 (which is size) - which makes me think we could write this as {entiresInDataBase, numberTo prune} ... ?

File: slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    final List<SignedAttestation> expectedAttestations = allAttestations.subList(pruningMark, size);
+    final List<SignedAttestation> attestations = getAttestations(1);
+    assertThat(attestations).usingFieldByFieldElementComparator().isEqualTo(expectedAttestations);
+
+    final List<SignedBlock> expectedBlocks = allBlocks.subList(pruningMark, size);
+    final List<SignedBlock> blocks = getBlocks(1);
+    assertThat(blocks).usingFieldByFieldElementComparator().isEqualTo(expectedBlocks);
+
+    final UInt64 expectedWatermarkValue = UInt64.valueOf(pruningMark);
+    assertThat(getWatermark(1))
```
Comment (rain-on) [inline slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java:50]: not sure why the 'getX' functions take a parameter, what does it mean? i.e. is it the validatorId? If so - could we call that out as a VALIDATOR_ID variable/constant?

File: slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/WatermarkUpdatesIntegrationTest.java [hunk 4]
```diff
@@ -23,7 +23,7 @@
 import org.apache.tuweni.units.bigints.UInt64;
 import org.junit.jupiter.api.Test;
 
-public class WatermarkUpdatesIntegrationTest extends InterchangeBaseIntegrationTest {
+public class WatermarkUpdatesIntegrationTest extends IntegrationTestBase {
```
Comment (rain-on) [inline slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/WatermarkUpdatesIntegrationTest.java:26]: that IntegrationTestBase renames could be a standalone PR, seems to have touched quite a bit (easy to review here, but touched many files)

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbLocker.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.web3signer.slashingprotection;
+
+import org.jdbi.v3.core.Handle;
+
+public class DbLocker {
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbLocker.java:17]: where'd this come from? Guessing we always had it ... but ... renamed?

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java [hunk 6]
```diff
@@ -110,6 +107,7 @@ public DbSlashingProtection(
                 .registerModule(new InterchangeModule())
                 .configure(FLUSH_AFTER_WRITE_VALUE, true)
                 .enable(SerializationFeature.INDENT_OUTPUT));
+    this.pruner = new Pruner(jdbi, signedBlocksDao, signedAttestationsDao, lowWatermarkDao);
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java:110]: sounds like Pruner should be a generic idea - and we should have a DB specific version of one

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    this.signedBlocksDao = signedBlocksDao;
+    this.signedAttestationsDao = signedAttestationsDao;
+    this.lowWatermarkDao = lowWatermarkDao;
+  }
+
+  public void pruneForValidator(
+      final int validatorId, final long epochsToKeep, final long slotsPerEpoch) {
+    checkArgument(
+        epochsToKeep > 0,
+        "epochsToKeep to keep must be a positive value, but was %s",
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:52]:
```suggestion
        "epochsToKeep must be a positive value, but was %s",
```

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        epochsToKeep);
+    checkArgument(
+        slotsPerEpoch > 0, "slotsPerEpoch must be a positive value, but was %", slotsPerEpoch);
+    final long slotsToKeep = Math.max(epochsToKeep / slotsPerEpoch, 1);
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:56]: are there slots in an epoch, or are there epochs in a slot?

coz the naming of the variables make me think this should be a multiple, not divide (i.e. should the variable be epochsPerSlot?)

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    pruneBlocks(validatorId, slotsToKeep);
+    pruneAttestations(validatorId, epochsToKeep);
+  }
+
+  private void pruneBlocks(final int validatorId, final long slotsToKeep) {
+    final Optional<UInt64> pruningSlot =
+        jdbi.inTransaction(
+            READ_UNCOMMITTED,
+            h -> {
+              lockForValidator(h, LockType.BLOCK, validatorId);
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:66]: is there any race condition between importing and Pruning that we hadn't thought of? I suspect its fine, as the watermark can only move upwards.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+              final Optional<UInt64> watermarkSlot =
+                  lowWatermarkDao
+                      .findLowWatermarkForValidator(h, validatorId)
+                      .map(SigningWatermark::getSlot);
+              final Optional<UInt64> slot = signedBlocksDao.findMaxSlot(h, validatorId);
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:71]: rename variable to "mostRecentSlot"? (mostRecentAttestationSlot, mostRecentAttestation)

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+      final Optional<UInt64> highpoint,
+      final Optional<UInt64> watermark) {
+    return highpoint.flatMap(
+        h ->
+            watermark.map(
+                w -> {
+                  final UInt64 pruningPoint =
+                      h.compareTo(UInt64.valueOf(amountToPrune)) < 0
+                          ? UInt64.ZERO
+                          : h.subtract(amountToPrune).add(1);
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:122]: why add 1?

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+              epochToPruneTo.ifPresent(
+                  e -> lowWatermarkDao.updateEpochWatermarksFor(h, validatorId, e, e));
+              return epochToPruneTo;
+            });
+
+    pruningEpoch.ifPresent(
+        e ->
+            jdbi.useTransaction(
+                READ_UNCOMMITTED,
+                h -> signedAttestationsDao.deleteAttestationsBelowEpoch(h, validatorId, e)));
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:108]: is it worth calling this "deleteAttestationsbelowWatermark"? no need to pass "e" in - and thus the signedAttestationsDao function becomes a bit safer... (but then has dependency on watermark :( )

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedAttestationsDao.java [hunk 13]
```diff
@@ -118,4 +118,20 @@ public void insertAttestation(final Handle handle, final SignedAttestation signe
                 + "FROM signed_attestations WHERE validator_id = ?")
         .bind(0, validatorId).mapToBean(SignedAttestation.class).stream();
   }
+
+  public void deleteAttestationsBelowEpoch(
+      final Handle handle, final int validatorId, final UInt64 epoch) {
+    handle.execute(
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedAttestationsDao.java:124]: feels like a really dangerous function to have openly available - i.e. what if epoch is above the watermark?

Comment (rain-on) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedAttestationsDao.java:124]: same goes for blocks

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
+/*
```
[the remark this answers is not in the document]
Comment (jframe) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java]: done

File: core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java [hunk 15]
```diff
@@ -68,36 +72,28 @@
...
     this.azureKeyVaultParameters = azureKeyVaultParameters;
   }
 
-  private Optional<SlashingProtection> createSlashingProtection(
-      final boolean slashingProtectionEnabled,
-      final String slashingProtectionDbUrl,
-      final String slashingProtectionDbUser,
-      final String slashingProtectionDbPassword) {
-    if (slashingProtectionEnabled) {
+  private Optional<SlashingProtection> createSlashingProtection() {
```
[the remark this answers is not in the document]
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java]: done

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2ExportSubCommand.java [hunk 16]
```diff
@@ -54,7 +54,7 @@ public void run() {
     if (output == null) {
       throw new MissingParameterException(
           spec.commandLine(), spec.findOption("--to"), "--to has not been specified");
-    } else if (StringUtils.isEmpty(eth2Config.slashingProtectionDbUrl)) {
+    } else if (StringUtils.isEmpty(eth2Config.slashingProtectionParameters.getDbUrl())) {
```
[the remark this answers is not in the document]
Comment (jframe) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/subcommands/Eth2ExportSubCommand.java]: done

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/signer/runner/CmdLineParamsDefaultImpl.java [hunk 17]
```diff
@@ -157,6 +157,17 @@ public CmdLineParamsDefaultImpl(final SignerConfiguration signerConfig, final Pa
...
     }
 
+    if (signerConfig.isSlashingProtectionPruningEnabled()) {
+      params.add("--slashing-protection-pruning-enabled");
+      params.add(Boolean.toString(signerConfig.isSlashingProtectionPruningEnabled()));
+      params.add("--slashing-protection-pruning-epochs-to-keep");
+      params.add(Long.toString(signerConfig.getSlashingProtectionPruningEpochsToKeep()));
+      params.add("--slashing-protection-pruning-slots-per-epoch");
+      params.add(Long.toString(signerConfig.getSlashingProtectionPruningSlotsPerEpoch()));
+      params.add("--slashing-protection-pruning-schedule");
```
[the remark this answers is not in the document]
Comment (jframe) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/signer/runner/CmdLineParamsDefaultImpl.java:167]: Thinking --slashing-protection-pruning-interval

File: acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/signer/SignerConfiguration.java [hunk 18]
```diff
@@ -185,6 +197,22 @@ public String getSlashingProtectionDbPassword() {
...
+
+  public long getSlashingProtectionPruningEpochsToKeep() {
+    return slashingPruningEpochsToKeep;
+  }
+
+  public long getSlashingProtectionPruningSlotsPerEpoch() {
+    return slashingPruningSlotsPerEpoch;
+  }
+
+  public long getSlashingProtectionPruningSchedule() {
```
[the remark this answers is not in the document]
Comment (jframe) [reply acceptance-tests/src/test/java/tech/pegasys/web3signer/dsl/signer/SignerConfiguration.java:212]: I'll add a subclass for slashing config

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+              + "(default: ${DEFAULT-VALUE})",
+      paramLabel = "<BOOL>",
+      arity = "1")
+  boolean pruningEnabled = false;
+
+  @Option(
+      names = {"--slashing-protection-pruning-epochs-to-keep"},
+      description = "Number of epochs to keep. (default: ${DEFAULT-VALUE})",
+      arity = "1")
+  long pruningEpochsToKeep = 10_000;
```
[the remark this answers is not in the document]
Comment (jframe) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java:62]: no idea. but for a single validator think this is fine.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+
+  @Option(
+      names = {"--slashing-protection-pruning-slots-per-epoch"},
+      description =
+          "Slots per epoch to use when calculating the number of slots to prune for signed"
+              + " blocks. This typically will not need changing and defaults to value used on mainnet "
+              + "(default: ${DEFAULT-VALUE})")
+  long pruningEpochsPerSlot = 32;
```
[the remark this answers is not in the document]
Comment (jframe) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java:70]: Prefer not to add another mixin at the moment. Don't there is a lot of value in doing this.

I'll rename pruningEpochsPerSlot to pruningSlotsPerEpoch

File: core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java [hunk 21]
```diff
@@ -111,6 +107,33 @@ protected String getOpenApiSpecResource() {
     return "openapi/web3signer-eth2.yaml";
   }
 
+  @Override
+  public void run() {
+    super.run();
+
+    if (slashingProtection.isPresent() && slashingProtectionParameters.isPruningEnabled()) {
```
[the remark this answers is not in the document]
Comment (jframe) [reply core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java:115]: I'll move this into the SlashingProtectionFactory

File: slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/WatermarkUpdatesIntegrationTest.java (hunk 4 above)
Comment (jframe) [reply slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/WatermarkUpdatesIntegrationTest.java]: done

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbLocker.java (hunk 5 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbLocker.java:17]: This was extracted from DbSlashingProtection

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java (hunk 6 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java:110]: it's only going to be used for the db slashing protection so don't think this makes sense to be generic at the moment

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java (hunk 9 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:66]: Think it is fine. We advise users not import while the service is running as well so this should also cover this.

File: slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+  void prunesDataForRegisteredValidator(final int amountToKeep, final int pruningMark) {
+    final int size = 10;
+    insertAndRegisterData(size, size, 1);
+    final List<SignedAttestation> allAttestations = getAttestations(1);
+    final List<SignedBlock> allBlocks = getBlocks(1);
+
+    slashingProtection.prune(amountToKeep, 1);
+
+    final List<SignedAttestation> expectedAttestations = allAttestations.subList(pruningMark, size);
+    final List<SignedAttestation> attestations = getAttestations(1);
```
[the remark this answers is not in the document]
Comment (jframe) [reply slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java:42]: done

File: slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java (hunk 2 above)
Comment (jframe) [reply slashing-protection/src/integration-test/java/tech/pegasys/web3signer/slashingprotection/PruningIntegrationTest.java:33]: renamed argument to expectedLowestPopulatedSlot, hopefully this is clearer

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java (hunk 7 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:52]: done

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java (hunk 8 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:56]: your right this should be a multiply not a divide. it is slots per epoch. i'll fix up the names in the config.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java (hunk 10 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:71]: done

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java (hunk 12 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java:108]: I like this, i'll update the DAO to do this. done

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedAttestationsDao.java (hunk 13 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedAttestationsDao.java:124]: Going to update this to only delete data below watermark

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java (hunk 11 above)
Comment (jframe) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java]: It's due the watermark we are calculating being inclusive of the amount of items to keep.

File: core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java [hunk 23]
```diff
@@ -194,8 +192,12 @@ private ArtifactSignerProvider loadSigners(
               }
 
               slashingProtection.ifPresent(
-                  slashingProtection -> slashingProtection.registerValidators(validators));
-
+                  slashingProtection -> {
```
Comment (rain-on) [inline core/src/main/java/tech/pegasys/web3signer/core/Eth2Runner.java:195]: raise a issue to check if we should prune JUST our data, or EVERYONE's data in the database?

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbPruner.java [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+    final Optional<UInt64> watermarkSlot =
+        lowWatermarkDao
+            .findLowWatermarkForValidator(handle, validatorId)
+            .map(SigningWatermark::getSlot);
+    final Optional<UInt64> mostRecentSlot = signedBlocksDao.findMaxSlot(handle, validatorId);
+    final Optional<UInt64> slotWatermark =
+        calculateWatermark(slotsToKeep, mostRecentSlot, watermarkSlot);
+    final Optional<SignedBlock> watermark =
+        slotWatermark.flatMap(
+            w -> signedBlocksDao.findNearestBlockWithSlot(handle, validatorId, w));
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbPruner.java:88]: nit: findFirstBlockNewerThanSlot(x)?

Comment (rain-on) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbPruner.java:88]: tbh - if slotWatermark doesn't exist, then no need to continue? [no need to change what's here, but just a thought]

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedBlocksDao.java [hunk 25]
```diff
@@ -70,4 +70,35 @@ public void insertBlockProposal(final Handle handle, final SignedBlock signedBlo
...
+
+  public Optional<UInt64> findMaxSlot(final Handle handle, final int validatorId) {
+    return handle
+        .createQuery("SELECT max(slot) FROM signed_blocks WHERE validator_id = ?")
+        .bind(0, validatorId)
+        .mapTo(UInt64.class)
+        .findFirst();
+  }
+
+  public Optional<SignedBlock> findNearestBlockWithSlot(
```
Comment (rain-on) [inline slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/dao/SignedBlocksDao.java:90]: findFirstBlockAfter(x)?

**Comment by jframe** [comment]
#371
